src/actions/actions.ts

```diff
@@ -0,0 +1,568 @@
+import { ModeHandler } from './../mode/modeHandler';
+import { ModeName } from './../mode/mode';
+import { showCmdLine } from './../cmd_line/main';
+import { Motion, MotionMode } from './../motion/motion';
+import {DeleteOperator} from './../operator/delete';
+import {ChangeOperator} from './../operator/change';
+import {PutOperator} from './../operator/put';
+import {TextEditor} from './../textEditor';
+import * as vscode from 'vscode';
+
+/**
+ * An action is the most basic sort of thing you can do in vimotion.
+ * 'h', 'j', 'w' etc are all actions.
+ */
+export abstract class BaseAction {
+  /**
+   * Modes that this action can be run in.
+   */
+  public modes: ModeName[];
+
+  /**
+   * The key you press to trigger the action.
+   */
+  public key: string;
+
+  /**
+   * Is this action valid in the current Vim state?
+   */
+  public doesActionApply(modeHandler: ModeHandler, key: string): boolean {
+    if (this.modes.indexOf(modeHandler.currentMode.name) === -1) { return false; }
+    if (this.key.indexOf(key) === -1) { return false; }
+
+    return true;
+  }
+
+  /**
+   * Run the action.
+   *
+   * TODO: The dream is to not pass in modeHandler, only motion.
+   * This is quite a far off dream, though.
+   */
+  public abstract async execAction(modeHandler: ModeHandler, motion: Motion): Promise<void>;
+}
+
+export class Actions {
+
+  /**
+   * Every Vim action will be added here with the @RegisterAction decorator.
+   */
+  public static allActions: BaseAction[] = [];
+}
+
+function RegisterAction(action) {
+    Actions.allActions.push(new action());
+}
+
+@RegisterAction
```

Comment:
Yeah, it's really snazzy. Unfortunately I'm running into some annoying bugs where depending on compile order things might not work, so I might have to remove it...